docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+* Provide an active-active example
+* Templating proposals need to be argued through to reduce options
+
+## Motivation
+
+Many examples of clustered software systems require stronger guarantees per instance than are provided
+by the Replication Controller (aka ReplicaSet). Instances of these systems typically require:
+
+1. Data per instance which should not be lost even if the pod is deleted, typically on a persistent volume
+   * Some cluster instances may have tens of TB of stored data - forcing new instances to replicate data
+     from other members over the network is onerous
+2. A stable and unique identity associated with that instance of the storage - such as a unique member id
+3. A consistent network identity that allows other members to locate the instance even if the pod is deleted
+4. A predictable number of instances to ensure that systems can form a quorum
+   * This may be necessary during initialization
+5. Instances can migrate from machine to machine as necessary and are not tied to an instance
+6. The ability to scale up in a controlled fashion, but are very rarely scaled down without human
+   intervention
+
+Kubernetes should expose a pod controller (a PetSet) that satisfies these requirements in a flexible
+manner. It should be easy for users to manage and reason about the behavior of this set. An administrator
+with familiarity in a particular cluster system should be able to leverage this controller and its
+supporting documentation to run that clustered system on Kubernetes. It is expected that some adaptation
+is required to support each new cluster.
+
+
+## Use Cases
+
+The software listed below forms the primary use-cases for a PetSet on the cluster - problems encountered
+while adapting these for Kubernetes should be addressed in a final design.
+
+* Quorum with Leader Election
+  * MongoDB - in replica set mode forms a quorum with an elected leader, but instances must be preconfigured
+    and have stable network identities.
+  * ZooKeeper - forms a quorum with an elected leader, but is sensitive to cluster membership changes and
+    replacement instances *must* present consistent identities
+  * etcd - forms a quorum with an elected leader, can alter cluster membership in a consistent way, and
+    requires stable network identities
+* Decentralized Quorum
+  * Cassandra - allows flexible consistency and distributes data via innate hash ring sharding, is also
+    flexible to scaling, more likely to support members that come and go. Scale down may trigger massive
+    rebalances.
+* Active-active
+  * **Need active-active master example - galera? **
+* ??? 
+
+
+## Background
+
+ReplicaSets are designed with a weak guarantee - that there should be N replicas of a particular pod template.
+Each pod instance varies only by name, and the replication controller errs on the side of ensuring that N
+replicas exist as quickly as possible (by creating new pods as soon as old ones begin graceful deletion, for
```

Comment:
Due to the assumption of fungibility, scaling down is opportunistic -- the controller has the freedom to pick arbitrary pods, such as ones that are not Ready, Pending, not well spread, etc.